### PR TITLE
Adding support to ingress-nginx as ingress controller with CNAME + TLS with Tsuru

### PR DESCRIPTION
**Pull request by gfleury**
- Add support to https://github.com/kubernetes/ingress-nginx
- Add support to TLS
- Add support to CNAMEs

Part of https://github.com/gfleury/tsuru-helm-chart

### CODE REVIEW

**Comment by andrestc** [comment]
This looks good, but I have a few considerations (I'll add some comments to the code). 

I dont think we need a separated type to handle nginx ingress. I'm thinking about replacing `https://github.com/tsuru/kubernetes-router/blob/master/kubernetes/ingress.go` with your implementation (as it seems more complete). There is not much to change except accepting `AnnotationsPrefix` and DefaultDomain as parameter to the struct. 
We could provide an nginx mode out of the box building up from this generic ingress service, in main.go, e.g:
```go
if mode == "ingressNginx" {
    k8sAnnotations["kubernetes.io/ingress.class"] = "nginx"
    base := &kubernetes.BaseService{
	Namespace:   *k8sNamespace,
	Timeout:     *k8sTimeout,
	Labels:      *k8sLabels,
	Annotations: *k8sAnnotations,
        AnnotationsPrefix: annotationsPrefix,
    }
    service = &kubernetes.IngressService{BaseService: base, DefaultDomain: defaultDomain}
}
```

Do you want to take a shot at this?

File: cmd/router/main.go [hunk 1]
```diff
@@ -33,6 +32,8 @@ func main() {
 	k8sAnnotations := &MapFlag{}
 	flag.Var(k8sAnnotations, "k8s-annotations", "Annotations to be added to each resource created. Expects KEY=VALUE format.")
 	ingressMode := flag.Bool("ingress-mode", false, "Creates ingress resources instead of LB services.")
+	ingressNginxMode := flag.Bool("ingressnginx-mode", false, "Creates ingress resources to use with ingress-nginx.")
```
Comment (andrestc) [inline cmd/router/main.go]: I think we could change this to a string flag, e.g, `--mode`, and validates that its one of the following: `service, ingress or ingressNginx`

File: kubernetes/ingress-nginx.go [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+// Get gets the address of the loadbalancer associated with
+// the app Ingress resource
+func (k *IngressNginxService) Get(appName string) (map[string]string, error) {
+	_, err := k.get(appName)
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]string{"address": fmt.Sprintf("%v.%v", appName, k.DefaultDomain)}, nil
```
Comment (andrestc) [inline kubernetes/ingress-nginx.go]: perhaps this should check if a custom domain was used?

File: kubernetes/ingress-nginx.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+func (k *IngressNginxService) secretClient() (typedV1.SecretInterface, error) {
+	client, err := k.getClient()
+	if err != nil {
+		return nil, err
+	}
+	return client.CoreV1().Secrets(k.Namespace), nil
+}
+
+func secretName(appName string) string {
+	return appName + "-secret"
```
Comment (andrestc) [inline kubernetes/ingress-nginx.go]: maybe adding `kubernetes-router` to this name would help avoid collisions and making the secret easier to recognize

**Comment by gfleury** [comment]
Yah, I can take a look on that.

**Comment by gfleury** [comment]
Tests still missing, will work on them tomorrow.

**Comment by gfleury** [comment]
I was on holidays, I will be working on the tests.

**Comment by gfleury** [comment]
@andrestc Hey, took almost a year but I couldn't work on these tests before today. Could you take a look?

ps: do you still own that black corsa? kkkkkkk cheers bro.

**Comment by andrestc** [comment]
hey @gfleury, this looks solid. I just need to do some testing on a real cluster to make sure everything is ok (probably by next week). Thanks for contributing this.

ps: I feel offended, its a prisma (and yes, I still own it hehehe) cheers!

**Pull request merged**